### PR TITLE
[SPARK-37176][SQL] Sync JsonInferSchema#infer method's exception handle logic with JacksonParser#parse method

### PR DESCRIPTION
**Pull request by advancedxy**
### What changes were proposed in this pull request?
Change `JsonInferSchema#infer`'s exception handle logic to be aligned with `JacksonParser#parse`

### Why are the changes needed?
To reduce behavior inconsistency, users can have the same expectation for schema infer and json parse when dealing with some malformed input.

### Does this PR introduce _any_ user-facing change?
Yes.
Before this patch, json's inferring schema could be failed for some malformed input but succeeded when parsing.
After this patch, they have the same exception handle logic.

### How was this patch tested?
Added one new test and modify one exist test to cover the new case.

### CODE REVIEW

**Comment by advancedxy** [comment]
Hi, @MaxGekk could you review this PR in your spare time since you touched the relevant(encoding related) code.

**Comment by MaxGekk** [comment]
@advancedxy Thank you for the ping. I will review this PR today.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala [hunk 1]
```diff
@@ -68,14 +83,18 @@ private[sql] class JsonInferSchema(options: JSONOptions) extends Serializable {
...
-            case PermissiveMode =>
-              Some(StructType(Seq(StructField(columnNameOfCorruptRecord, StringType))))
-            case DropMalformedMode =>
-              None
-            case FailFastMode =>
-              throw QueryExecutionErrors.malformedRecordsDetectedInSchemaInferenceError(e)
-          }
+          case e: SparkUpgradeException => throw e
+          case e @ (_: RuntimeException | _: JsonProcessingException |
+                    _: MalformedInputException) =>
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:88]: Could you write tests or modify existing tests to check new exception handling:
- SparkUpgradeException
- MalformedInputException

Comment (advancedxy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala:88]: Good point.  Will add later.

**Comment by HyukjinKwon** [comment]
im okay with this change. I think this is dupe of https://github.com/apache/spark/pull/33706 but let's go ahead with this PR. cc @yaooqinn FYI

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala (hunk 1 above)
Comment (advancedxy) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala]: The `SparkUpgradeException` is not necessary, so I removed this exception handling.
I add a new test to cover `MalformedInputException`

**Comment by advancedxy** [comment]
> im okay with this change. I think this is dupe of #33706 but let's go ahead with this PR. cc @yaooqinn FYI


Ah, I didn't see that one. @yaooqinn would like to finish your PR? If so, you can copied some of my tests into your commit.  Otherwise, I will proceed with PR.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@advancedxy Your PR has better test coverage for now. I am waiting for test results and going to merge it if @HyukjinKwon @yaooqinn don't mind.

**Comment by yaooqinn** [comment]
I am OK, this change looks good to me

**Review by yaooqinn — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank youm @advancedxy and @yaooqinn @HyukjinKwon for review.

**Pull request closed without merging**